### PR TITLE
fix sync_proxy_rules_iptables_total metric

### LINKED ISSUE #118978: meaning of sync_proxy_rules_iptables_total given MinimizeIPTablesRestore
The `sync_proxy_rules_iptables_total` metric is defined as "Number of proxy iptables rules programmed".

In the past, that effectively meant "the total number of rules in iptables that kube-proxy owns". But post-`MinimizeIPTablesRestore`, it now ends up being "the number of iptables rules that changed in the last sync". (FWIW this is only a change for the `nat` half of the metric; currently we always rewrite all `filter` rules, even in a partial sync, so the `filter` half of the metric is still "correct".)

There are arguments for either meaning, but I think the traditional ("total number of rules in iptables that kube-proxy owns") is more useful?

One possibility would be to only set the metric during full syncs, but in that case we might go a very long time without updating it, and it would get out of date.

Another possibility would be that instead of setting that metric from the `iptables-restore` _input_, we could set it from the `iptables-save` _output_. This would mean that it would always lag by (at least) one sync (since it would be set based on the rules we are replacing rather than the rules we are creating) but that would at least be more up-to-date than waiting for the next full resync. (In `largeClusterMode`, it would lag a bit more, only getting updated once every `syncPeriod` (default 30s).)

If we think it's important to have it always be up-to-date then we could try to figure out some way to calculate the right value during a partial update... in that case we'd have to run through the code that currently gets bypassed during a partial update, and have it decide what rules are needed, but not actually try to write them.

/sig network

### PR DESCRIPTION
**Pull request by danwinship**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Reverts the definition of the `sync_proxy_rules_iptables_total` metric back to the generally-understood pre-`MinimizeIPTablesRestore` meaning: "the total number of iptables rules that kube-proxy is responsible for". Also adds a new metric, `sync_proxy_rules_iptables_last`, preserving the behavior that `sync_proxy_rules_iptables_total` had accidentally slipped into: "the number of iptables rules that kube-proxy reprogrammed on the last sync".

Also fixes a bug noticed while added unit tests for this, which is that if syncProxyRules() deleted any stale service/endpoint chains, it would count each of those _deletions_ as being a "rule" for purposes of the metric due to carelessness in how it was counting.

#### Which issue(s) this PR fixes:
Fixes #118978

#### Does this PR introduce a user-facing change?
```release-note
The kube-proxy `sync_proxy_rules_iptables_total` metric has now reverted back
to its pre-1.27 behavior of tracking the total number of iptables rules that
kube-proxy is responsible for, rather than only counting the number of rules
that it re-synced on the last sync. The new `sync_proxy_rules_iptables_last`
metric now gives the latter number.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3453-minimize-iptables-restore/README.md
```

/sig network
/priority important-soon
/assign @thockin @aojea

### CODE REVIEW

**Comment by aojea** [comment]
LGTM

the only doubts I have is about the solution of switching the pointers to store the lines in a different buffer, it is not straightforward for future developments to realize there is some code doing that in the middle of the loop.
There is no problem if the future rules have to be skipped though

/assign @thockin

**Comment by danwinship** [comment]
Yeah, I didn't love that approach, but it seemed simplest...

I guess I could try doing it with separate pointers throughout, like `requiredNATRules` and `fullSyncOnlyNATRules`, and it would decide at the top (when it sets `tryPartialSync` whether to set `fullSyncOnlyNATRules` to the same buffer as `requiredNATRules` or a dummy buffer.

(This would actually make one of the earlier refactorings (#110266) irrelevant; I'd carefully reorganized all of the code in the main sync loop so that the required rules all come first, and then the rules that can be skipped at the bottom. So maybe if we refactored it with `requiredNATRules` vs `fullSyncOnlyNATRules`, I should undo some of that reorganization, because maybe it would be more logical to go back to grouping things the old way if we don't have to worry about being able to skip the second half of the loop?)

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -852,8 +852,8 @@ func (proxier *Proxier) syncProxyRules() {
 	proxier.natChains.Reset()
 	proxier.natRules.Reset()
 
-	skippedNatChains := &proxyutil.LineBuffer{}
-	skippedNatRules := &proxyutil.LineBuffer{}
+	skippedNatChains := proxyutil.NewDummyLineBuffer()
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go:855]: A comment here would not hurt

File: pkg/proxy/util/linebuffer.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	return buf.lines
+}
+
+type dummyLineBuffer struct {
+	lines int
+}
+
+// NewDummyLineBuffer returns a dummy LineBuffer that counts the number of writes but
+// throws away the data.
+func NewDummyLineBuffer() LineBuffer {
```
Comment (thockin) [inline pkg/proxy/util/linebuffer.go:117]: Maybe `NewLineCounter()` or `NewDiscardLineBuffer()`  would make it feel like less of a test-infra thing?

**Review by thockin — COMMENTED** [review]
I don't hate the pointer flip.  I wonder if we can insure against accidentally writing to `proxier.natRules` by making some of these helper methods into free-functions, but that can be a followup, I think.

**Comment by danwinship** [comment]
updated the name
/hold cancel

**Pull request merged**